### PR TITLE
Room switching

### PR DESCRIPTION
**Pull request by frederikcreemers**
I've implemented room switching using react-router. This is probably a less fluxy way to do it than keeping the current room/channel in a store, but it makes the URL the single source of truth for the current channel.

You'll see a lot of `componentWillReceiveProps` added. This is because url params are passed to RouteHandler components by the router, but the components aren't necessarily unmounted and mounted again when the url changes.

### CODE REVIEW

File: src/components/chat/chat.jsx [hunk 1]
```diff
@@ -8,8 +8,19 @@ const ChatChannels = require('./chat_channels.jsx');
...
+
 
 const Chat = React.createClass({
+  mixins: [Router.State],
+  componentWillReceiveProps() {
+    this.setState({currentRoom: this.getParams().roomSlug});
+  },
+
+  getInitialState(){
+      return {currentRoom:this.getParams().roomSlug};
```
Comment (pletcher) [inline src/components/chat/chat.jsx]: Style nit:

```
return {
  currentRoom: this.getParams().roomSlug
};
```

would be preferable.

File: src/components/chat/chat_channels.jsx [hunk 2]
```diff
@@ -39,6 +43,10 @@ const ChatChannels = React.createClass({
     UsersStore.removeChangeListener(this.updateUsers);
   },
 
+  componentWillReceiveProps() {
+      this.setState({currentChannel: this.getParams().roomSlug});
```
Comment (pletcher) [inline src/components/chat/chat_channels.jsx]: Same style nit as above.

File: src/components/chat/chat_channels.jsx [hunk 3]
```diff
@@ -51,7 +59,7 @@ const ChatChannels = React.createClass({
 
   getInitialState() {
     return {
-      currentChannel: ChatChannelsStore.getCurrentChannel(),
+      currentChannel: this.getParams().roomSlug,
```
Comment (pletcher) [inline src/components/chat/chat_channels.jsx:64]: Does `getParams()` return at least an empty object (`{}`)? I just want to make sure we don't trigger a `cannot get property "roomSlug" of undefined` error if `getParams()` returns undefined.

File: src/components/chat/chat_channels.jsx [hunk 4]
```diff
@@ -84,10 +92,9 @@ const ChatChannels = React.createClass({
   renderChannels() {
     return this.state.channels.map((channel) => {
       let label = channel.slug;
-      let url = `/rooms/${label}`;
 
       return (
-        <a className="block white px3 h5 light-gray" href={url} key={url}>#{label}</a>
+        <Link to="chat" params={{roomSlug: label}} key={label} className="block white px3 h5 light-gray">#{label}</Link>
```
Comment (pletcher) [inline src/components/chat/chat_channels.jsx:99]: :+1:

File: src/stores/chat_channels_store.js [hunk 5]
```diff
@@ -17,9 +17,6 @@ class ChatChannelsStore extends Store {
         case ActionTypes.CHAT_CHANNELS_RECEIVED:
           channels = List(action.channels);
           break;
-        case ActionTypes.CHAT_CHANNEL_RECEIVED:
```
Comment (pletcher) [inline src/stores/chat_channels_store.js:20]: Nice.

File: src/mixins/current_channel_mixin.js [hunk 6]
```diff
@@ -1,23 +0,0 @@
-'use strict';
```
Comment (pletcher) [inline src/mixins/current_channel_mixin.js:1]: :tada:

**Comment by pletcher** [comment]
@bigblind, just a couple of style nits -- then let's get tests passing (I'll get Travis set up in the meantime) and merge this in.

File: src/components/chat/chat_channels.jsx (hunk 3 above)
Comment (frederikcreemers) [inline src/components/chat/chat_channels.jsx:64]: @pletcher yes it does, and roomSlug will always be present, otherwise the route wouldn't match.

Comment (pletcher) [inline src/components/chat/chat_channels.jsx:64]: Sweet! React Router is new-ish to me, so just making sure I have a good handle on its ins and outs.

**Comment by pletcher** [comment]
I'll fix the merge conflicts -- that was my bad.

**Pull request merged**